File: model_1/CS3233_2024_Midterm/key_item_recovery_1_solution.py
n = int(input().strip())
matrix = []
for _ in range(n):
    matrix.append(list(map(int, input().split())))
edges = []
for i in range(2, n + 1):
    row = matrix[i - 1]
    M = 0
    for j in range(1, i):
        if row[j - 1] > M:
            M = row[j - 1]
    u = min(i, M)
    v = max(i, M)
    edges.append((u, v))
edges.sort(key=lambda x: (-x[0], x[1]))
for u, v in edges:
    print(f"{u} {v}")

File: model_1/CS3233_2024_Midterm/key_item_recovery_1_solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int n;
    cin >> n;
    vector<vector<int>> matrix(n, vector<int>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> matrix[i][j];
        }
    }

    vector<pair<int, int>> edges;

    for (int i = 2; i <= n; i++) {
        int row_index = i - 1;
        int parent = 0;
        for (int j = 0; j < i - 1; j++) {
            if (matrix[row_index][j] > parent) {
                parent = matrix[row_index][j];
            }
        }
        edges.push_back(make_pair(parent, i));
    }

    sort(edges.begin(), edges.end(), [](const pair<int, int>& a, const pair<int, int>& b) {
        if (a.first != b.first) {
            return a.first > b.first;
        }
        return a.second < b.second;
    });

    for (auto& edge : edges) {
        cout << edge.first << " " << edge.second << endl;
    }

    return 0;
}